tests/approaches/test_grammar_search_invention_approach.py

```diff
@@ -487,3 +490,48 @@ def test_branching_factor_heuristic():
     # is lower than we would like. These are actually the predicates that get
     # returned by running the grammar search on covers with branching factor.
     assert forall_not_covers_h < holding_h
+
+
+def test_task_planning_heuristic():
+    """Tests for _TaskPlanningHeuristic().
+    """
+    # We know that this heuristic is bad, because it's way too optimistic: it
+    # thinks that any valid sequence of operators can be refined into a plan.
+    # This unit test illustrates that pitfall.
+    utils.update_config({
+        "env": "cover",
+    })
+    utils.update_config({
+        "env": "cover",
+        "offline_data_method": "demo+replay",
+        "seed": 0,
+    })
+    env = CoverEnv()
+
+    name_to_pred = {p.name : p for p in env.predicates}
+    Holding = name_to_pred["Holding"]
+    HandEmpty = name_to_pred["HandEmpty"]
+
+    candidates = {
+        Holding: 1.0,
+        HandEmpty: 1.0,
+    }
+    train_tasks = next(env.train_tasks_generator())
+    dataset = create_dataset(env, train_tasks)
+    atom_dataset = utils.create_ground_atom_dataset(dataset,
+        env.goal_predicates | set(candidates))
+    heuristic = _TaskPlanningHeuristic(env.goal_predicates, atom_dataset,
+                                       train_tasks, candidates)
+    all_included_h = heuristic.evaluate({Holding, HandEmpty})
+    none_included_h = heuristic.evaluate(set())
+    # This is terrible!
```

Comment:
Why exactly is this happening? Is it because the operators with no added predicates just think you can always achieve Covers, which works at the task planning level?

Reply:
yeah, exactly